### DISCUSSION TITLE
Is OPC UA / IEC104 bidirectional

### DISCUSSION THREAD

**Discussion by iavor-k** [riclolsen/json-scada#172] [Q&A]
Hello, I’m currently building a bidirectional OPC UA to IEC 104 gateway. When I configure tags on the IEC 104 client connection, they become available through the OPC UA server as expected. This is absolutely great! 
My question is: Does JSON-SCADA automatically issue a write command to the IEC 104 driver when a value is written via OPC UA, or it requires manually sending commands through the MongoDB collection?

**Comment by riclolsen** [comment]
Hi! Thanks for the feedback!
No, this is not automatic. Routing a command requires you say how it should be routed. Do this 
- create a command tag 
- configure its protocol source parameters for the IEC 104 connection
- add a protocol destination with parameters for the command you want at the OPC-UA connection

This is confusing because commands flow in the opposite direction as monitored points. You must notice that in this case the source is 104 and destination is OPC. So, when the command is received on the OPC connection it will be translated to IEC 104 accordingly to the parameters configured.
Best regards,
Ricardo

**Reply by iavor-k** [reply]
Hello Ricardo, 
From what I understand, I need to set the IEC104 Client connection as the source protocol for my command tag. Then I have to add the  OPC UA Server connection as destination protocol. However, when I try to add a destination from the "Admin/Tags/Edit Tag/Add destination" page, nothing shows up in the list of options. I'm attaching the mongo db collections with my test configuration:

[json_scada.protocolConnections.json](https://github.com/user-attachments/files/17967494/json_scada.protocolConnections.json)
[json_scada.protocolDriverInstances.json](https://github.com/user-attachments/files/17967495/json_scada.protocolDriverInstances.json)
[json_scada.realtimeData.json](https://github.com/user-attachments/files/17967497/json_scada.realtimeData.json)

Do you think I am missing something in the setup process?
Thanks again for your assistance!
Best regards,
Iavor

**Reply by riclolsen** [reply]
Yup, you discovered a bug, thanks.
Please just replace the src/AdminUI/dist folder contents and it should work.
I will update the source code in the repo ASAP.
[dist.zip](https://github.com/user-attachments/files/17969161/dist.zip)

**Comment by iavor-k** [comment]
Hi again,
I replaced the "dist" folder but no destination protocol options appeared on the screen. As I was not sure if the web app was actually updated (due to my limited docker knowledge) , I pulled the new version from the repo (including commit d596b68a0d4759589649cc3c892ac79b3860a154) and start everything from scratch. The result was the same. Then I reloaded the original demo project and added new IEC104 connection and a new tag. I noticed that when I try to add the destination protocol only IEC104 DIST is available in the drop down list. It seems like the app filters the connections suggested for destination by some criteria and the OPC-UA DIST doesn't match. This could explain why in the beginning I didn't see any items in the list - I had no IEC104 DIST connection in my project. I checked the OPC-UA DIST options and I can confirm that the commands are enabled. Is there any hidden property which makes the protocol visible in the destination list?
Thanks!

**Comment by riclolsen** [comment]
Indeed there was still something missing.
OPCUAServer were not listed as destinations for tags. Also some fields must accept text and not just numbers.
I have updated the code.
As you are using docker, you can do a git pull and run the docker compile again.
When configuring the destination put the tag name of the command in the object address field. In the ASDU TI type put the OPC type, "boolean" for digital commands or "double" for analog commands. Other fields leave at default value.
